src/data/list/defs.lean

```diff
@@ -170,13 +170,88 @@ def find (p : α → Prop) [decidable_pred p] : list α → option α
 def mfind {α} {m : Type → Type u} [monad m] [alternative m] (tac : α → m unit) : list α → m α :=
 list.mfirst $ λ a, tac a $> a
 
-def find_indexes_aux (p : α → Prop) [decidable_pred p] : list α → nat → list nat
-| []     n := []
-| (a::l) n := let t := find_indexes_aux l (succ n) in if p a then n :: t else t
+/-- Auxiliary definition for `foldl_with_index`. -/
+def foldl_with_index_aux (f : ℕ → α → β → α) : ℕ → α → list β → α
+| _ a [] := a
+| i a (b :: l) := foldl_with_index_aux (i + 1) (f i a b) l
+
+/-- Fold a list from left to right as with `foldl`, but the combining function
+also receives each element's index. -/
+def foldl_with_index (f : ℕ → α → β → α) (a : α) (l : list β) : α :=
+foldl_with_index_aux f 0 a l
+
+/-- Auxiliary definition for `foldr_with_index`. -/
+def foldr_with_index_aux (f : ℕ → α → β → β) : ℕ → β → list α → β
+| _ b [] := b
+| i b (a :: l) := f i a (foldr_with_index_aux (i + 1) b l)
+
+/-- Fold a list from right to left as with `foldr`, but the combining function
+also receives each element's index. -/
+def foldr_with_index (f : ℕ → α → β → β) (b : β) (l : list α) : β :=
+foldr_with_index_aux f 0 b l
+
+/-- `indexes l` is the list of indexes of `l`, i.e. `[0 .. l.length - 1]`. -/
+def indexes (l : list α) : list ℕ :=
+l.foldr_with_index (λ i _ is, i :: is) []
+
+/-- `indexed l` is `l` but every element is annotated with its index.
+
+     indexed l = zip (indexes l) l -/
+def indexed (l : list α) : list (ℕ × α) :=
+l.foldr_with_index (λ i a out, (i, a) :: out) []
 
 /-- `find_indexes p l` is the list of indexes of elements of `l` that satisfy `p`. -/
 def find_indexes (p : α → Prop) [decidable_pred p] (l : list α) : list nat :=
-find_indexes_aux p l 0
+foldr_with_index (λ i a is, if p a then i :: is else is) [] l
+
+/-- Returns the elements of `l` that satisfy `p` together with their indexes in
+`l`. The returned list is ordered by index. -/
+def indexes_values (p : α → Prop) [decidable_pred p] (l : list α) : list (ℕ × α) :=
+foldr_with_index (λ i a l, if p a then (i , a) :: l else l) [] l
+
+/-- `indexes_of a l` is the list of all indexes of `a` in `l`.
+
+     indexes_of a [a, b, a, a] = [0, 2, 3] -/
```

Comment:
Add backticks, and maybe say something like "For example".

Reply:
(I realise these problems predate your PR!)

Reply:
Backticks shouldn't be necessary since the indentation already causes these to be rendered as code blocks in VS Code and on the doc page.

Reply:
What Bryan says. But Scott is now the second maintainer to get confused by this, so I've changed it anyway.